### PR TITLE
Add support for ERC20 tokens

### LINKED ISSUE #21: Add support for ERC20 token

### PR DESCRIPTION
**Pull request by QEDK**
- [X] Overhauls contract flow to only use ERC20 tokens
- [x] Add support in frontend for approvals

Fixes #21

### CODE REVIEW

**Comment by Dhaiwat10** [comment]
Added a temp modal that asks for stake amount to recruiters when they click on the request information button. The `request` and `approveRequest` functionality is working fine.

One problem: For some reason, now I cannot unlock my own information. Well, I _can_ unlock it but I can't _see_ it. The backend returns an error. This was possible earlier. I did both the transactions successfully - request and approve request. Not sure if that's by design but just putting it out here.

Other notes: Had to rebase this branch against staging since this branch didn't have the latest frontend code.

cc @QEDK @Cali93

File: packages/react-app/components/cards/MediaCard.jsx [hunk 1]
```diff
@@ -70,27 +72,55 @@ function MediaCard({
   const toast = useToast();
 
   const handleRequestPrivateProfileUnlock = async () => {
+    const stakeAmount = prompt("Enter stake amount in MATIC");
```
Comment (QEDK) [inline packages/react-app/components/cards/MediaCard.jsx]: Replace this with Chakra's modal: https://chakra-ui.com/docs/overlay/modal

File: packages/react-app/components/cards/MediaCard.jsx [hunk 2]
```diff
@@ -70,27 +72,55 @@ function MediaCard({
...
+    if (!stakeAmount) {
+      return alert("Please enter a valid stake amount");
```
Comment (QEDK) [inline packages/react-app/components/cards/MediaCard.jsx]: Perform validation using the NumberInput component: https://chakra-ui.com/docs/form/number-input

File: packages/react-app/components/cards/MediaCard.jsx [hunk 3]
```diff
@@ -70,27 +72,55 @@ function MediaCard({
...
+    }
+    const weiStakeAmount = ethers.utils.parseEther(stakeAmount);
     try {
-      const tx = await dRecruitContract.request(privateProfile.tokenId, {
-        value: ethers.utils.parseEther("0.01"),
+      // Request allowance
+      const dERC20Address = await dRecruitContract.token();
+      const signer = await context.injectedProvider.getSigner();
+      const dERC20Contract = new ethers.Contract(dERC20Address, abis.ERC20, signer);
```
Comment (QEDK) [inline packages/react-app/components/cards/MediaCard.jsx]: This process should not happen on the fly but stored in state on first load, similar to other contracts. You can put it as the last step of `loadWeb3Modal()`

File: packages/react-app/components/cards/MediaCard.jsx [hunk 4]
```diff
@@ -70,27 +72,55 @@ function MediaCard({
...
+      const allowance = await dERC20Contract.allowance(context.address, dRecruitContract.address);
+      // Only ask for allowance if it is not enough
+      if (allowance < weiStakeAmount) {
```
Comment (QEDK) [inline packages/react-app/components/cards/MediaCard.jsx]: Also check if allowance is below the threshold amount, 1 * 10 ** 18, we don't want dupe allowance txs

**Comment by Dhaiwat10** [comment]
Hey mate, thanks for reviewing. I don't have access to a laptop right now -
kinda an emergency, but I'll make these changes tomorrow. Dropping this msg
since I haven't been able to get back to this all day

File: packages/react-app/components/cards/MediaCard.jsx [hunk 5]
```diff
@@ -69,140 +83,199 @@ function MediaCard({
...
+    }
+    const weiStakeAmount = ethers.utils.parseEther(stakeAmount);
     try {
-      const tx = await dRecruitContract.request(privateProfile.tokenId, {
-        value: ethers.utils.parseEther("0.01"),
+      // Request allowance
+      const allowance = await tokenContract.allowance(context.address, dRecruitContract.address);
+      // Only ask for allowance if it is not enough
+      if (allowance.lt(weiStakeAmount) || allowance.lt(ethers.utils.parseEther("0.1"))) {
+        await tokenContract.approve(dRecruitContract.address, weiStakeAmount);
```
Comment (QEDK) [inline packages/react-app/components/cards/MediaCard.jsx]: Store this as a tx object and await it with `.wait()` We need to wait until the tx gets mined, not until it's sent.

File: packages/react-app/components/cards/MediaCard.jsx [hunk 6]
```diff
@@ -69,140 +83,199 @@ function MediaCard({
...
+            />
+          </ModalBody>
+
+          <ModalFooter>
+            <Button
+              marginLeft={0}
+              marginRight={"auto"}
+              colorScheme="blue"
+              onClick={() => {
+                onClose();
```
Comment (QEDK) [inline packages/react-app/components/cards/MediaCard.jsx]: Don't close the modal on transaction being sent! We need the context to exist, close the modal after.

Comment (Dhaiwat10) [reply packages/react-app/components/cards/MediaCard.jsx]: Bit confused here. The modal doesn't send a transaction. It just captures the stake amount and stores it in a state variable

File: packages/react-app/components/cards/MediaCard.jsx (hunk 5 above)
Comment (Dhaiwat10) [reply packages/react-app/components/cards/MediaCard.jsx]: Done

File: packages/react-app/components/cards/MediaCard.jsx (hunk 6 above)
Comment (QEDK) [reply packages/react-app/components/cards/MediaCard.jsx]: It's a confusing UX pattern for transaction prompts to appear after the modal closes. Functionally, I understand the intent but for an user, it's jarring.

File: packages/react-app/components/cards/MediaCard.jsx [hunk 7]
```diff
@@ -69,140 +83,201 @@ function MediaCard({
...
+      const allowance = await tokenContract.allowance(context.address, dRecruitContract.address);
+      // Only ask for allowance if it is not enough
+      if (allowance.lt(weiStakeAmount) || allowance.lt(ethers.utils.parseEther("0.1"))) {
+        const tx = await tokenContract.approve(dRecruitContract.address, weiStakeAmount);
+        await tx.wait();
+      }
+      const tx = await dRecruitContract.request(privateProfile.tokenId, weiStakeAmount, {
+        value: 0,
       });
+      await tx.wait();
```
Comment (QEDK) [inline packages/react-app/components/cards/MediaCard.jsx]: This is handled below.

File: packages/react-app/components/cards/MediaCard.jsx (hunk 6 above)
Comment (Dhaiwat10) [reply packages/react-app/components/cards/MediaCard.jsx]: I see. I'll work on this next. Won't be able to work on it for dogfooding tonight tho :(

**Pull request merged**